### PR TITLE
tmpfile writer to solve rsync problem on windows - issue #42

### PR DESCRIPTION
**Pull request by gfilardo**
Hi pstadler,
I forked your project to address the rsync problem on windows reported on #42. 
To achieve the cross-platform compatibility, I had to change the path of the file used by rsync listing all the files to transfer, since rsync does not work with path containing the drive letter followed by colon.
On unix-like systems the behavior is completely unaltered. On the other hand, on windows, the cwd is used to contain the temporary file.
I wrote the new `temp-file-write` module, which is similar to `temp-write`, but allows you to specify the destination directory (the os temp dir is used by default). I used it in place of the original one in the `shell.js` file, together with the `os` module to recognize the platform.
The system works fine, but I would like to highlight two potential problems:
- on windows the process using flightplan has to have writing capabilities on the current working directory
- if a problem should arise during the transfer, the tmp file would not be deleted

That being said, I would like to state that in my work I have tested the system and I haven't encountered such problems, using both `deltacopy` and `cwRsync`. There are three major requirements on windows, though: `git` has to be  in the path, as well as `python` 2, and fibers have to be supported. An installation of Visual Studio 2013 provides all the required dlls to support them.

### CODE REVIEW

**Comment by pstadler** [comment]
Hi Giovanni

I'd like to ask you to rename `lib/temp-file/temp-file-write.js` to `lib/utils/index.js` instead and to move the OS check inside the writer function, so that the interface is clean:
```var writeTempFile = require('./utils').writeTempFile
var tmpFile = writeTempFile(files);
```

Besides that, I don't see why we need graceful-fs?

Please let me know if you're tired of doing this, so I will do the changes followed by a single commit, which I want to avoid, because you did the work.

Thanks!

**Comment by gfilardo** [comment]
Hi Patrick,
I will refactor the temp file writer as you suggested tomorrow.
In answer to your two questions:

- given the very specific nature of the refactored temp file writer, the graceful-fs module should not be needed anymore, I think the dependency could be safely removed in this case, but I would prefer to do some tests on my windows computers at work on Monday
- I think that pyhton is required by node-gyp which is in turn required by node-fibers

Expect the changes done and tested by Monday. 
Thank you

**Comment by gfilardo** [comment]
Hi Patrick,
I have refactored the temp file writer as you asked and successfully tested it on mac and on windows.

File: lib/utils/index.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+
+function tempFilePath(filepath) {
+  return path.join(filepath || tmpdir, uuid.v1());
+}
+
+module.exports.writeTempFile = function (str) {
+  var isWin = /^win/i.test(os.platform());
+  var fullpath = isWin ? tempFilePath('.') : tempFilePath();
+
+  mkdirp.sync(path.dirname(fullpath));
```
Comment (pstadler) [inline lib/utils/index.js]: I think it's not necessary to create folders at either `.` or whatever `require('os').tmpdir()` returns, as they already exist.

**Comment by pstadler** [comment]
Thanks man, please see my comment above.

File: lib/utils/index.js [hunk 2]
```diff
@@ -0,0 +1,20 @@
+var path = require('path')
+  , tmpdir = require('os').tmpdir()
```
Comment (pstadler) [inline lib/utils/index.js]: please move this below line 5, and use `tmpdir = os.tmpdir()`

**Comment by gfilardo** [comment]
Done!

**Comment by pstadler** [comment]
Thanks man, this is great.

One last task, could you squash this into a single commit?

Cheers
Patrick

**Comment by pstadler** [comment]
One more thing, sorry for being so picky, but having less dependencies is a good thing.
`uuid` could be replaced with something like this:

```javascript
require('crypto').randomBytes(16).toString('hex')
```

**Comment by gfilardo** [comment]
Sure, I will push the changes tomorrow morning.

**Comment by pstadler** [comment]
Looks good, can you rebase / squash this into a single commit?

**Comment by gfilardo** [comment]
Done!

**Comment by pstadler** [comment]
Thanks man, will try to land a new version this weekend!

**Comment by pstadler** [comment]
Release notes: https://github.com/pstadler/flightplan/releases/tag/0.6.2

**Pull request merged**